### DISCUSSION TITLE
ViewLocator not working

### DISCUSSION THREAD

**Discussion by AscendLiu** [AvaloniaUI/Avalonia#17566] [Q&A]
Hello, is there anyone who can guide you on how to troubleshoot when ViewLocator is not working? Debug found that only some controls are triggered, while others that are needed are not
```
    <Application.DataTemplates>
        <avalonia:ViewLocator />
    </Application.DataTemplates>

```
![image](https://github.com/user-attachments/assets/54f45504-ad4b-4a02-a330-f07ab4f11e60)

**Comment by stevemonaco** [comment]
You're debugging the correct method, so every data that bubbles up to the root ViewLocator (not handled by other `DataTemplate`s you might have) will go through that method. If you're mixing ReactiveUI and Mvvm Toolkit in the same project, you'll want to ensure that you're matching on both. ie. If `ViewModelBase` is a `ReactiveObject`.

If that's not sufficient, then I suggest adding some logging (even if it's just `Debug.WriteLine`) to the Match method to ensure that the expected VMs are reaching that far.

**Reply by AscendLiu** [reply]
First of all, thank you for your help. I roughly understand what you mean. The data I called must have been processed by my custom or other Data Templates. The key point is, I don't understand when this Data Template was called? How should I debug, as this DataTemplates call seems to occur internally, and my code logic is very simple

**Reply by thevortexcloud** [reply]
When Avalonia sees a control it does not have a template for, it searches up the tree until it finds one. Debugging it is generally just a matter of making sure your matching and instantiation logic is right.

**Reply by AscendLiu** [reply]
I am a beginner, and this is explained in the documentation, but I don't know how to operate it specifically

**Comment by timunie** [comment]
You may want to check Avalonia.Samples to learn more about DataTemplates, which the ViewLocator is in the end.

**Reply by thevortexcloud** [reply]
Here is the sample: https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/DataTemplates/

**Comment by AscendLiu** [comment]
I'm very sorry for taking up everyone's time. Thank you again. However, I have seen all the examples on this official website. Creating a project named ViewLocator on my own also works, but I am currently using it in conjunction with a third-party library. I don't know why it doesn't work, and I can't solve it because I don't know how to debug and find out the reason why it doesn't work and triggers it
![image](https://github.com/user-attachments/assets/f5438564-3c0e-46c0-a7ca-0c662d0e3afd)


**Actually, what I want to ask is, is there a way to directly debug and find the moment when ViewLocator should be called**

**Reply by stevemonaco** [reply]
In general, not from application code. Locating a matching `DataTemplate` (of which ViewLocator is one) can require an upwards traversal of the entire logical tree of controls in your app. You can check the source here: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/Templates/DataTemplateExtensions.cs

Presenters (not seen in your XAML, but part of many controls like `ContentControl`) are used to present both controls and data. You're presenting data here, so we'll focus to that.

Presenters can explicitly be provided a `ContentTemplate`, which is what you're doing above, so data without visuals can be "presented". This has highest priority because you explicitly assign it. If you did not provide one, then Avalonia tries to locate and match `DataTemplate`s at each level upwards until it finds a match or reaches your ViewLocator at the app root (in this case, since you're using one).

**Reply by thevortexcloud** [reply]
I seem to recall there is some weirdness as well when it comes to using styles from another assembly. No idea if it applies here though.

https://docs.avaloniaui.net/docs/guides/styles-and-resources/how-to-use-included-styles

**Comment by AscendLiu** [comment]
Thank you again for your help. I have looked at the source code and have a better understanding of the Data Templates in the documentation, but I still cannot find the reason. Possible reason: Styles, but I have explicitly added the third-party library's Style file, but I have excluded it;
I used  **DevTools**  to browse VisualTree and found that the generated template is indeed incorrect. I suspect that it is due to the higher priority of internal Data Templates, which is indeed a design principle.

However, with almost identical code, I changed the files related to the ViewModelt and Views, and the effect improved again. I know the path is incorrect, and ViewLocator may not match correctly, but my current situation is that the incorrect path prevents ViewLocator from being entered at all.
So, **I don't know if the placement path of files in my own project will cause different priority matching**, or if there are other reasons, but I have temporarily solved it this way.
